### PR TITLE
Show how to use Dune to compile examples

### LINKED ISSUE #506: Document update using dune as primary build system?
https://ocaml.org/docs/modules
The document is still base on ocamlopt , I guess it a bit out of sync modern OCaml world.
Now we using dune for every single , small or large project of OCaml.
So the question is , why not just teach user using dune as primary build system for learning OCaml.

I tried to learn OCaml to build with ocamlopt, but it hard to manage a real size OCaml project.
Dune looks a real good tools , just like maven to java, cmake to cpp/c, cargo to rust.
So please rewrite document, base on dune build tools.

### PR DESCRIPTION
**Pull request by cuihtlauac**
Fix: https://github.com/ocaml/ocaml.org/issues/506

Usage of ocamlc and ocamlopt remains. Text is added showing how
to achieve the same result using dune with a minimum set up.

### CODE REVIEW

**Comment by tmattio** [comment]
This is great, thanks!

Generally, though, I think we should avoid recommending ocamlc and ocamlopt as the default, and reserve their usage only to "advanced" how to guides. Having a unique recommendation (dune) throughout the doc will make the onboarding experience much less confusing to newcomers.

For this specific case, and since we don't have such guides yet, I'd recommend moving the `dune exec` command above, and putting the `ocamlc/ocamlopt` section below, as the alternative.

File: data/tutorials/lg_03_modules.md [hunk 1]
```diff
@@ -11,6 +11,8 @@ date: 2021-05-27T21:07:30-00:00
 
 ## Basic usage
 
+### Files defines modules
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
### Files Define Modules
```

File: data/tutorials/lg_03_modules.md [hunk 2]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
 let () = Amodule.hello ()
 ```
 
-We can compile the files in one command:
+### Automatized compilation
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
### Automatized Compilation
```

File: data/tutorials/lg_03_modules.md [hunk 3]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+
+In order to compile them using the [Dune](https://dune.build/) build system,
+which is now the standard on OCaml, at least two configuration files are
+required:
+
+* File `dune-project` which contains project-wide configuration data. Here very
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
* The `dune-project` file, which contains project-wide configuration data. Here's very
```

File: data/tutorials/lg_03_modules.md [hunk 4]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+  minimal one:
+  ```
+   (lang dune 3.4)
+  ```
+* File `dune` which contains actual build directives, a project may have several
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
* The `dune` file, which contains actual build directives. A project may have several
```

File: data/tutorials/lg_03_modules.md [hunk 5]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+  of them, depending on the organization of the source. This is sufficient for
+  this example:
+  ```
+  (executable (name bmodule))
+  ```
+
+Here is how to create the configuration files, build the source and run the
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
Here is how to create the configuration files, build the source, and run the
```

File: data/tutorials/lg_03_modules.md [hunk 6]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+<!-- $MDX dir=examples -->
+```bash
+$ echo "(lang dune 3.4)" > dune-project
+$ echo "(executable (name bmodule))" > dune
+$ dune build
+$ dune exec ./bmodule.exe
+Hello
+```
+
+Actually, `dune build` is optional, simply running `dune exec` would have
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
Actually, `dune build` is optional. Simply running `dune exec` would have
```

File: data/tutorials/lg_03_modules.md [hunk 7]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+triggered the compilation. Beware in the `dune exec` command, the parameter
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
triggered the compilation. Beware in the `dune exec` command, as the parameter
```

File: data/tutorials/lg_03_modules.md [hunk 8]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+`./bmodule.exe` is not a file path. This command means “execute the content of
+the file `./bmodule.ml`”. However, the actual executable file is stored and
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
the file `./bmodule.ml`." However, the actual executable file is stored and
```

File: data/tutorials/lg_03_modules.md [hunk 9]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+named differently.
+
+In a real world project, it is preferable to start by creating the Dune
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
In a real world project, it is preferable to start by creating the `dune`
```

File: data/tutorials/lg_03_modules.md [hunk 10]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+configuration files and directory structure using the `dune init project`
+command.
+
+### Manual compilation
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
### Manual Compilation
```

File: data/tutorials/lg_03_modules.md [hunk 11]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+
+Alternatively, it is possible, but not recommended, to compile the files by
+directly calling the compiler, using a single command:
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
directly calling the compiler by using a single command:
```

File: data/tutorials/lg_03_modules.md [hunk 12]
```diff
@@ -54,6 +98,15 @@ $ ocamlopt -c bmodule.ml
...
 ```
 
+In both case, a stand alone executable is created
+<!-- $MDX dir=examples -->
+```sh
+$ ./hello
+Hello
+```
+
+### Naming and scoping
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
### Naming and Scoping
```

File: data/tutorials/lg_03_modules.md [hunk 13]
```diff
@@ -54,6 +98,15 @@ $ ocamlopt -c bmodule.ml
...
+
 Now we have an executable that prints "Hello". As you can see, if you want to
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
Now we have an executable that prints "Hello." As you can see, if you want to
```

File: data/tutorials/lg_03_modules.md [hunk 14]
```diff
@@ -117,21 +170,21 @@ to something called `amodule2.ml`:
...
 ```ocaml
 val message : string
-
 val hello : unit -> unit
 ```
 
 Let's assume that accessing the `message` value directly is none of the others
-modules' business. We want to hide it by defining a restricted interface. This
-is our `amodule2.mli` file:
+modules' business, we want it to be a private definition. We can hide it by
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
modules' business; we want it to be a private definition. We can hide it by
```

File: data/tutorials/lg_03_modules.md [hunk 15]
```diff
@@ -143,14 +196,41 @@ val hello : unit -> unit
 to document .mli files using the format supported by
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
to document `.mli` files using the format supported by
```

File: data/tutorials/lg_03_modules.md [hunk 16]
```diff
@@ -143,14 +196,41 @@ val hello : unit -> unit
...
-compiled using `ocamlc`, even if .ml files are compiled to native code using
-`ocamlopt`:
+The corresponding module `Bmodule2` is defined in file `bmodule2.ml`:
+
+<!-- $MDX file=examples/bmodule2.ml -->
+```ocaml
+let () = Amodule2.hello ()
+```
+
+The .mli files must be compiled before the matching .ml files. This is done
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
The .`mli` files must be compiled before the matching `.ml` files. This is done
```

File: data/tutorials/lg_03_modules.md [hunk 17]
```diff
@@ -143,14 +196,41 @@ val hello : unit -> unit
...
+$ echo "(executables (names bmodule bmodule2))" > dune
+$ dune build
+$ dune exec ./bmodule.exe
+Hello
+$ dune exec ./bmodule2.exe
+Hello 2
+```
+
+There how the same result can be acheived by calling the compiler manually.
+Notice the .mli file is compiled using byte-code compiler `ocamlc` , while if
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
Notice the `.mli` file is compiled using byte-code compiler `ocamlc` , while if
```

File: data/tutorials/lg_03_modules.md [hunk 18]
```diff
@@ -143,14 +196,41 @@ val hello : unit -> unit
...
+.ml files are compiled to native code using `ocamlopt`:
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
`.ml` files are compiled to native code using `ocamlopt`:
```

File: data/tutorials/lg_03_modules.md [hunk 19]
```diff
@@ -143,14 +196,41 @@ val hello : unit -> unit
...
 $ ocamlopt -c amodule2.ml
+$ ocamlopt -c bmodule2.ml
+$ ocamlopt -o hello2 amodule2.cmx bmodule2.cmx
+$ ./hello
+Hello
+$ ./hello2
+Hello 2
 ```
 
 ## Abstract types
```
Comment (christinerose) [inline data/tutorials/lg_03_modules.md]:
```suggestion
## Abstract Types
```

**Review by christinerose — CHANGES_REQUESTED** [review]
Just some minor grammar/formatting suggestions. I could go through the rest of the file after you merge if you'd like (and open my own PR). Let me know if you'd like me to do that. :)

File: data/tutorials/lg_03_modules.md [hunk 20]
```diff
@@ -34,7 +36,49 @@ And here is what we have in `bmodule.ml`:
...
+$ echo "(executable (name bmodule))" > dune
+$ dune build
+$ dune exec ./bmodule.exe
+Hello
+```
+
+Actually, `dune build` is optional. Simply running `dune exec` would have
+triggered the compilation. Beware in the `dune exec` command, as the parameter
+`./bmodule.exe` is not a file path. This command means “execute the content of
+the file `./bmodule.ml`”. However, the actual executable file is stored and
```
Comment (cuihtlauac) [inline data/tutorials/lg_03_modules.md]:
```suggestion
the file `./bmodule.ml`.” However, the actual executable file is stored and
```

File: data/tutorials/lg_03_modules.md [hunk 21]
```diff
@@ -54,6 +98,15 @@ $ ocamlopt -c bmodule.ml
...
+In both case, a stand alone executable is created
+<!-- $MDX dir=examples -->
+```sh
+$ ./hello
+Hello
+```
+
+### Naming and Scoping
+
 Now we have an executable that prints "Hello". As you can see, if you want to
```
Comment (cuihtlauac) [inline data/tutorials/lg_03_modules.md]:
```suggestion
Now we have an executable that prints `Hello`. As you can see, if you want to
```

**Review by cuihtlauac — COMMENTED** [review]
Issue acknowledged. Fix updated

**Review by tmattio — APPROVED** [review]

**Pull request merged**